core/src/main/scala/org/apache/spark/shuffle/ShuffleBlockPusher.scala

```diff
@@ -0,0 +1,456 @@
+  private[this] val pushRequests = new Queue[PushRequest]
+  private[this] val errorHandler = createErrorHandler()
+  // VisibleForTesting
+  private[shuffle] val unreachableBlockMgrs = new HashSet[BlockManagerId]()
+  private[this] var stopPushing = false
+
+  // VisibleForTesting
+  private[shuffle] def createErrorHandler(): BlockPushErrorHandler = {
+    new BlockPushErrorHandler() {
+      // For a connection exception against a particular host, we will stop pushing any
+      // blocks to just that host and continue push blocks to other hosts. So, here push of
+      // all blocks will only stop when it is "Too Late". Also see updateStateAndCheckIfPushMore.
+      override def shouldRetryError(t: Throwable): Boolean = {
+        // If the block is too late, there is no need to retry it
+        !Throwables.getStackTraceAsString(t).contains(BlockPushErrorHandler.TOO_LATE_MESSAGE_SUFFIX)
+      }
+    }
+  }
+
+  /**
+   * Initiates the block push.
+   *
+   * @param dataFile         mapper generated shuffle data file
+   * @param partitionLengths array of shuffle block size so we can tell shuffle block
+   * @param dep              shuffle dependency to get shuffle ID and the location of remote shuffle
+   *                         services to push local shuffle blocks
+   * @param mapIndex      map index of the shuffle map task
+   */
+  private[shuffle] def initiateBlockPush(
+      dataFile: File,
+      partitionLengths: Array[Long],
+      dep: ShuffleDependency[_, _, _],
+      mapIndex: Int): Unit = {
+    val numPartitions = dep.partitioner.numPartitions
+    val transportConf = SparkTransportConf.fromSparkConf(conf, "shuffle")
+    val requests = prepareBlockPushRequests(numPartitions, mapIndex, dep.shuffleId, dataFile,
+      partitionLengths, dep.getMergerLocs, transportConf)
+    // Randomize the orders of the PushRequest, so different mappers pushing blocks at the same
+    // time won't be pushing the same ranges of shuffle partitions.
+    pushRequests ++= Utils.randomize(requests)
+
+    submitTask(() => {
+      pushUpToMax()
+    })
+  }
+
+  /**
+   * Triggers the push. It's a separate method for testing.
+   * VisibleForTesting
+   */
+  protected def submitTask(task: Runnable): Unit = {
+    if (BLOCK_PUSHER_POOL != null) {
+      BLOCK_PUSHER_POOL.execute(task)
+    }
+  }
+
+  /**
+   * Since multiple block push threads could potentially be calling pushUpToMax for the same
+   * mapper, we synchronize access to this method so that only one thread can push blocks for
+   * a given mapper. This helps to simplify access to the shared states. The down side of this
+   * is that we could unnecessarily block other mappers' block pushes if all the threads
+   * are occupied by block pushes from the same mapper.
+   *
+   * This code is similar to ShuffleBlockFetcherIterator#fetchUpToMaxBytes in how it throttles
+   * the data transfer between shuffle client/server.
+   */
+  private def pushUpToMax(): Unit = synchronized {
```

Comment:
Check `stopPushing` at the beginning?

Reply:
If `stopPushing` is `false` this method will never be called. The first time this method is run is in `initiateBlockPush` and after that it is always submitted after a response from the server is received (line  204 - 206)
```         
 if (updateStateAndCheckIfPushMore(
            sizeMap(result.blockId), address, remainingBlocks, result)) {
            pushUpToMax()
          }
```
Here, `updateStateAndCheckIfPushMore` will return false when it has to stop pushing.

Reply:
The reason to check for `stopPushing` in the while loops here because it is possible that while the thread is processing these requests from the queue, there is an already  `Too Late` response received. Even though no more `pushUpToMax` tasks are submitted, these while loop will continue sending requests. I discovered this with the UT that I added `More blocks are not pushed when a block push fails with too late exception`

Reply:
> The reason to check for stopPushing in the while loops here because it is possible that while the thread is processing these requests from the queue, there is an already Too Late response received.....

`updateStateAndCheckIfPushMore()` and `pushUpToMax()` require the same lock. So they should be called synchronously. So I'm not sure how does the issue in UT you mentioned happens.

Reply:
Yes, that's right. `pushUpToMax` will be completed before `updateStateAndCheckIfPushMore()` is called. So, this happens only in the unit test, because for testing `submitTask`  is made to run synchronously and response of a block push is also handled by the same thread synchronously. I will remove this flag then and figure out a different way for the test.

Reply:
What I meant is that in the unit tests `updateStateAndCheckIfPushMore` can be called before `pushUpToMax` is completed. The UT is using a single thread and in `pushUpToMax`, `sendRequest` will invoke `mockBlockStoreClient.pushBlocks` which in turn will call `handleResult`.

Reply:
I have removed the `stopPushing` flag which I just added for the UT and instead fixed the tests.

Reply:
I see.